File: lib/commands/storyboard_remove/domain/storyboard_remove_usecase.dart
import 'package:icli/di/di.dart';
import 'package:icli/utils/file_helper.dart';
import 'package:icli/utils/resources.dart';
import 'package:icli/utils/ui.dart';

class StoryBoardRemoveUseCase {
  UI? ui;
  Resources? resources;
  FileHelper? fileHelper;

  StoryBoardRemoveUseCase(
      {UI? ui, FileHelper? fileHelper, Resources? resources}) {
    this.ui = get<UI>(object: ui);
    this.fileHelper = get<FileHelper>(object: fileHelper);
    this.resources = get<Resources>(object: resources);
  }

  run() {
    final xcodeproj = fileHelper?.find(RegExp(r'.*\.xcodeproj'), '.');
    final projectName = xcodeproj?.replaceFirst('.xcodeproj', '');
    final pbxPath = "$xcodeproj/project.pbxproj";

    if (projectName != null) {
      _removeFiles(projectName, pbxPath);
      _overrideInfoplist(projectName);
      _overrideAppDelegate(projectName);
    } else {
      ui?.error("Cound not find main project dir");
    }
  }

  _removeFiles(String projectDir, String pbxPath) {
    final files = [
      "$projectDir/Base.lproj/Main.storyboard",
      "$projectDir/SceneDelegate.swift"
    ];

    fileHelper?.rm(files, removeFromPbx: pbxPath);
  }

  _overrideInfoplist(String projectDir) {
    final info = resources?.infoplist();
    final old = projectDir + "/Info.plist";

    if (info != null) {
      fileHelper?.cp(info, old, true);
      ui?.echo("overwritten Info.plist", Color.green);
    } else {
      ui?.error("could not get Info.plist template");
    }
  }

  _overrideAppDelegate(String projectDir) {
    final info = resources?.appDelegate();
    final old = projectDir + "/AppDelegate.swift";

    if (info != null) {
      fileHelper?.cp(info, old, true);
      ui?.echo("overwritten AppDelegate.swift", Color.green);
    } else {
      ui?.error("could not get AppDelegate template");
    }
  }
}

File: icli/lib/usecase/storyboard_remove/storyboard_remove_usecase.rb
# frozen_string_literal: true

module ICLI
  # Removes storyboard structure from Xcode new project scaffold.
  # It will remove some files, modifies info.plist and pbx.
  class StoryboardRemoveUsecase
    def initialize
      @ui = ServiceLocator.resolve UI
      @project = ServiceLocator.resolve Project
      @resources = ServiceLocator.resolve Resources
      @file_helper = ServiceLocator.resolve FileHelper

      @main = 'Base.lproj/Main.storyboard'
      @scene = 'SceneDelegate.swift'
      @plist = 'Info.plist'
      @delegate = 'AppDelegate.swift'
    end

    def run
      proj = @project.open('.')
      target_path = proj.targets.first

      files_map = {
        @main => "#{target_path}/#{@main}",
        @scene => "#{target_path}/#{@scene}",
        @plist => "#{target_path}/#{@plist}",
        @delegate => "#{target_path}/#{@delegate}"
      }

      # remove_files files_map
      # TODO: wip need to remove storyboard reference from pbx info
      remove_files_from_pbx proj, files_map
      # overwrite_app_delegate files_map
      # overwrite_app_infoplist files_map
    end

    private

    def remove_files(files_map)
      files_to_remove = [files_map[@main], files_map[@scene]]
      @file_helper.rm paths: files_to_remove
    end

    def overwrite_app_delegate(files_map)
      @file_helper.cp from: @resources.app_delegate, to: files_map[@delegate]
    end

    def overwrite_app_infoplist(files_map)
      @file_helper.cp from: @resources.infoplist, to: files_map[@plist]
    end

    def find_files_in_project
      @file_helper.find_files '.', @files_to_remove
    end

    def remove_files_from_pbx(proj, files_map)
      get_files_ref(proj, files_map).each { |f| @project.rm_file f }

      @project.save proj
    end

    def get_files_ref(proj, files_map)
      path_prefix = @project.absolute_path(proj)

      [
        @project.file_ref(proj, "#{path_prefix}/#{files_map[@scene]}"),
        @project.file_ref(proj, "#{path_prefix}/#{files_map[@main]}").parent
      ]
    end
  end
end
